### PR TITLE
Merge entries in duplicate detection

### PR DESCRIPTION
**Pull request by oscargus**
Added a "merged entry" tab and preview/textbox in the DuplicateResolverDialog and added handling of this at import and duplicate checking.

Also fixed a bug leading to that if import was done without the ImportInspectionDialog, nullPointerException happened as the MetaData was not properly initialized.

### CODE REVIEW

**Comment by oscargus** [comment]
I just now realized that one can access this from the ImportInspectionDialog. Will sort that out first before there is any need to consider this...

**Comment by koppor** [comment]
When calling "Search" -> "Find duplicates", I cannot access this feature, but I really need that: Typically, I want to merge duplicates and not killing one of it.

**Comment by oscargus** [comment]
Did the "Possible duplicate entries" only contain two entries and nothing like below?

jabrefduplicate

**Comment by oscargus** [comment]
Anyway, I've updated the PR to also handle the ImportInspectionDialog. Some minor things to correct still, like using descriptions that make sense.

**Comment by koppor** [comment]
Oh, I see. In my case, it were only two entries. What are you doing when there are three entries? Merge the first two and then again showing the duplicate dialog for the merged entry and the third dublicate? (Repeat if there are more than 3)

**Comment by koppor** [comment]
Why are commits by me appearing here? Did you do any git magic? Maybe you should only do `git merge` and no `rebase -i`, `push -f` or `reset branch to here` :innocent:

**Comment by oscargus** [comment]
I'm only showing the new dialog... Not sure how JabRef handles more than two entries, but I do not think there were support for it earlier either?

Not sure, I thought it was the lack of all that magic that made them appear... :-) Now, there should be fewer though.

**Comment by oscargus** [comment]
No... No idea why your commits showed up originally. I thought I had a grip of those things now... 

I'll try to create a clean patch eventually by copy-and-paste.

**Comment by koppor** [comment]
@oscargus With some magic, you can create a new branch X based on master, cherry-pick commits (get the id via `gitk --all` and use the command line `git cherry-pick ID`), and then do `git push -f X origin:mergeduplicates`. Then, the branch of this PR is clean again.

Assumption: Your github upstream repo is called `origin`.

**Comment by oscargus** [comment]
OK! I'll see what I can come up with... The idea of cherry-picking instead of copy-pasted struck me at about the same time that you wrote your comment. :-)

**Comment by oscargus** [comment]
OK, I might have done this in a slightly different way, but still the end result is quite OK. Note that there is a "cleaner" PR in #228 .

**Comment by oscargus** [comment]
Oh, now ALL the old commits disappeared. :-) I close PR #228 then. By the way, for future reference the magic command was `git push -f origin X:mergeduplicates`.

**Comment by koppor** [comment]
The UI is a bit confusing.

Could you label the upper one with "first entry", the "lower" one with "second" entry and change the labels of the buttons accordingly?

When reading "Keep lower", I think, I pick the third entry.

Can there be a "Skip" button added? What if I don't want to merge the entry, but continue the searching? Usecase: An entry which isn't a duplicate.

File: src/main/java/net/sf/jabref/importer/ImportMenuItem.java [hunk 1]
```diff
@@ -219,34 +224,66 @@ public void update() {
                             }
                         }
 
+                        mainLoop:
```
Comment (simonharrer) [inline src/main/java/net/sf/jabref/importer/ImportMenuItem.java]: can you please solve this without such labels? they really complicate maintainability.

File: src/main/java/net/sf/jabref/importer/ImportMenuItem.java [hunk 2]
```diff
@@ -50,6 +50,7 @@
  */
 public class ImportMenuItem extends JMenuItem implements ActionListener {
 
+    private static final long serialVersionUID = 6164143717906802116L;
```
Comment (simonharrer) [inline src/main/java/net/sf/jabref/importer/ImportMenuItem.java]: these IDs are everywhere ;-)

File: src/main/java/net/sf/jabref/gui/JabRefFrame.java [hunk 3]
```diff
@@ -1927,16 +1928,22 @@ private int addBibEntries(List<BibtexEntry> bibentries, String filename,
...
+                            LOGGER.error("Merge button!");
+                            dupli = true;
+                            remove = true;
+                            toAdd.add(drd.getMergedEntry());
                         } else if (res == DuplicateResolverDialog.BREAK) {
                             break mainLoop;
                         }
+                        if (remove) {
+                            database.removeEntry(existingEntry.getId());
+                            ce.addEdit(new UndoableRemoveEntry(database, existingEntry, basePanel));
```
Comment (simonharrer) [inline src/main/java/net/sf/jabref/gui/JabRefFrame.java]: why are entries removed directly from the DB, but the addition of entries is delayed until after the loop?

File: src/main/java/net/sf/jabref/importer/ImportMenuItem.java (hunk 1 above)
Comment (oscargus) [inline src/main/java/net/sf/jabref/importer/ImportMenuItem.java]: Hmm, I simply stole that piece of code from another JabRef source file... But, yes, probably.

File: src/main/java/net/sf/jabref/importer/ImportMenuItem.java (hunk 2 above)
Comment (oscargus) [inline src/main/java/net/sf/jabref/importer/ImportMenuItem.java]: At least I removed them from the files I edited today. :-) (Or so I tried.)

**Comment by simonharrer** [comment]
Hm, would it make sense to integrate the dialog for merging entries with the duplication search?

File: src/main/java/net/sf/jabref/gui/JabRefFrame.java (hunk 3 above)
Comment (oscargus) [inline src/main/java/net/sf/jabref/gui/JabRefFrame.java]: Not sure. Just kept the old behavior, but indeed it makes sense to delay it.

I should for sure have removed my print-debugging on line 1933 though...

**Comment by oscargus** [comment]
Not sure if I fully understand the idea. At the moment you either know that you want to merge two entries (or import data from DOI for a single entry) and get the MergeEntryDialog or you rely on the duplication detection and get the DuplicateResolverDialog. I'm not sure that you always want to merge entries, so sometimes it is quicker just to drop one of them.

**Comment by simonharrer** [comment]
Okay. Last remark: what about removing the short form and the complete record tab? The merged entry tab shows all the relevant information including a simplistic form of highlighting the differences per field while both other tabs do not (in this case one has to find the differences without guidance).

**Comment by oscargus** [comment]
To me that makes sense, but let us hear any possible objections before I
remove the old stuff. I rarely use duplicate detection anyway and so far
mainly merge by hand.

One may think of an additional button in the MergeEntries panel for giving
priority, i.e. "clicking" all radio buttons on that side, to either the
left or the right entry in a single click (now the left entry has priority).

(Configurable in preferences? ;-))

(I also see a quite general problem here in my PRs. I mainly add
functionality rather than rethinking the design.)

**Comment by simonharrer** [comment]
Please, no additional preferences. ;) We try to keep this set as small as possible.

I was thinking to have the Keep left/Keep right buttons still. 

But we could still add two buttons for the merge entries to select all left or all right elements. Maybe make the left/right label into a clickable button?

**Comment by koppor** [comment]
:+1: for only "merged entry" without any additional tabs.

:+1: for replacing "Keep entry 1" and "Keep entry 2" by buttons nearby "Use ... Left" and "Use ... Right". Maybe a radio button directly below "Left" and one radio button directly below "Right".

**Comment by simonharrer** [comment]
I would keep the buttons for keep left\right. Because it saves the user a
click.
Am 21.10.2015 4:47 nachm. schrieb "Oliver Kopp" <notifications@github.com>:

**Comment by oscargus** [comment]
The idea was not to remove any buttons, just add the "preference button". I do not think a radio button conceptually makes sense there, but a regular button to select all left/right radio buttons (and add any missing fields from the other entry) for later editing.

The "entry 1"/"entry 2" problem is solved naturally with only this "tab" as left/right now makes much more sense.

Related question: when invoked from the ImportInspectionDialog, it states (approx)  "In database"/"From import". Would it make sense to keep some sort of similar behavior? (I think it does.)

@koppor Regarding your "Skip" button. The "Keep both" button does the same thing now.

Good thing: The DuplicateResolverDialog will disappear and only the MergeEntriesDialog will remain.

**Comment by oscargus** [comment]
Draft version without tabs and possibly correct naming of headings and buttons. Still a separate file though (realized that was by far the easiest).

Still to do:

- [ ] get rid of mainLoop
- [ ] postpone the entry removal
- [ ] add left/right buttons (for preferring fields from left or right entry)

**Comment by oscargus** [comment]
Looking at the code there is a preference to use the ImportInspectionDialog or not when importing from a file (first lines in the General preference tab). Any objections against always using the  ImportInspectionDialog? The alternative (if not checking this in the preferences) is to import them directly, but trigger the DuplicateResolvingDialog for every duplicate, which is a bit more tedious for the user.

Benefits: 2/3 preference items removed. 100+ lines of code removed.

**Comment by simonharrer** [comment]
Hm, hard to decide. I never use the import feature at all.

**Comment by oscargus** [comment]
I spent some more time thinking about it (the latest commit contains the result) and I think it makes sense. The alternative behavior can still be obtained by just clicking OK in the ImportInspectionDialog and then run duplicate detection.

The most likely use case I guess is to merge the contents of two .bib-files and there it should be more manageable to see the overview provided in the the ImportInspectionDialog instead of ending up in an indefinite (or at least hard to predict the length of) sequence of duplicate detections. And as said above, no actual functionality is lost, the alternative way just require one click and one menu selection.

It did solve two of the problems listed as well as removed duplicated code.

**Comment by simonharrer** [comment]
Sounds reasonable. Good job!

I just tried the "find duplicate" feature on your branch: the window is always quite small. When I increase the window size for the first duplication pair, it will be shrinked to the default size for the next duplication pair. This is a little bit annoying.

**Comment by oscargus** [comment]
I agree. I'll see if I can find some suitable code to copy...

**Comment by simonharrer** [comment]
for me this looks ready to be merged.

**Comment by oscargus** [comment]
I think the only missing thing that we have discussed (as far as I recall)
is to provide a way to simply select all left or right radio buttons.
Didn't really have the focus to sort it out (in a nice looking way) and may
not be really required to merge at this stage.

**Comment by koppor** [comment]
@oscargus Sure. I just wonder about the typo and the commented out code. Could you fix it or comment why these are needed?

**Comment by simonharrer** [comment]
:+1: for merging in, even without the mark left or right side options as we still have the keep left and keep right button.

**Comment by oscargus** [comment]
Fixed the two things pointed out by @koppor

**Comment by simonharrer** [comment]
Thank you for this PR again!

**Pull request merged**